File: Data Structures & Algorithms/valid-parenthesis-string/submission-3.cpp
class Solution {
public:
    bool checkValidString(string s) {
        std::stack<int> ast{};
        std::stack<int> paren{};
        for (auto i{0}; i < s.size(); ++i) {
            if (s[i] == '*') {
                ast.push(i);
            } else if (s[i] == '(') {
                paren.push(i);
            } else {
                if (paren.size() != 0) {
                    paren.pop();
                } else {
                    if (ast.size() == 0) {
                        return false;
                    } else {
                        ast.pop();
                    }
                }
            }
        }
        while (paren.size() > 0) {
            if (ast.size() == 0) {
                return false;
            } else if (paren.top() > ast.top()) {
                return false;
            } else {
                paren.pop();
                ast.pop();
            }
        }
        return true;
    }
};

File: Data Structures & Algorithms/minimum-stack/submission-3.py
class MinStack:

    def __init__(self):
        self.stack = []
        self.pref = []

    def push(self, val: int) -> None:
        self.stack.append(val)
        if len(self.pref):
            self.pref.append(min(self.pref[-1], val))
        else:
            self.pref.append(val)
    def pop(self) -> None:
        self.stack.pop()
        self.pref.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.pref[-1]
